src/stream.rs

```diff
@@ -0,0 +1,517 @@
+        } else {
+            // The requested block was already written to the file.
+            // Look for an existing mapping of the block.
+            let existing_mmap = self.mappings.get(&block_base);
+
+            // If there is no existing mapping, create one.
+            let mmap = match existing_mmap {
+                Some(mmap) => Arc::clone(mmap),
+                None => {
+                    // Get the file handle to be used for mapping.
+                    // The writer sets this before writing the first block.
+                    let file_guard = self.shared.file.load();
+                    let file_option = file_guard.deref().as_ref();
+                    let file_arc = file_option.unwrap();
+                    let file = file_arc.deref();
+                    // The block was already written to the file and will not
+                    // be modified by the writer, so it is safe to map it.
+                    let mmap_result = unsafe {
+                        MmapOptions::new()
+                            .offset(block_base)
+                            .len(Self::block_size())
+                            .map(file)
+                    };
+                    let new_mmap =
+                        Arc::new(mmap_result.map_err(StreamError::MapFile)?);
+                    self.mappings.push(block_base, Arc::clone(&new_mmap));
+                    new_mmap
+                }
+            };
+
+            // Return a handle to access the data through this mapping.
+            Ok(Mapped(mmap, range_in_block))
+        }
+    }
+
+    /// Block size in bytes.
+    pub const fn block_size() -> usize {
+        BLOCK_SIZE
+    }
+
+    /// Bitmask for bits defining an offset within a block.
+    pub const fn block_mask() -> usize {
+        BLOCK_SIZE - 1
+    }
+}
+
+impl<const BLOCK_SIZE: usize> Buffer<BLOCK_SIZE> {
+    /// Create a new buffer for the specified block.
+    fn new(block_base: u64)-> Result<Self, StreamError>{
+        Ok(Buffer {
+            block_base,
+            // Calling System.alloc safely requires the layout to be known to
+            // have non-zero size. If the allocation fails we return an error.
+            // Otherwise, our Drop impl guarantees the allocation is freed.
+            ptr: unsafe {
+                let ptr = System.alloc(Self::block_layout());
+                if ptr.is_null() {
+                    return Err(StreamError::Alloc());
+                }
+                ptr
+            }
+        })
+    }
+
+    /// Layout for allocating block-sized, block-aligned chunks of memory.
+    const fn block_layout() -> Layout {
+        assert!(BLOCK_SIZE != 0);
+        assert!(BLOCK_SIZE.is_power_of_two());
+        assert!(BLOCK_SIZE < (isize::MAX as usize));
+        // Safety: Align is non-zero, a power of 2, and does not overflow isize.
+        unsafe {
+            Layout::from_size_align_unchecked(BLOCK_SIZE, BLOCK_SIZE)
+        }
```

Comment:
If we're willing to require Rust 1.57 (which stabilised const_panic), this unsafe block could be replaced with something like:
```rust
        match Layout::from_size_align(BLOCK_SIZE, BLOCK_SIZE) {
            Ok(x) => x,
            _ => panic!("..."),
        }
```
The only major distro I can see that's too old is current Debian stable (bullseye), but bookworm should be out very soon.
Otherwise, we could leave it as a TODO for the future.

Reply:
With this PR merged we'll require at least 1.51 for const generics, which already exceeds the 1.48 in bullseye. I suspect that our MSRV may already be above that in practice. I've been building with 1.63 for quite a while, which is what's going to be in bookworm. I'm not sure we actually require it for anything yet, but I've not been checking that. How do the versions look on the other distros you looked at - what would be a sensible MSRV at this point?

> ```rust
> match Layout::from_size_align(BLOCK_SIZE, BLOCK_SIZE) {
>     Ok(x) => x,
>     _ => panic!("..."),
>    }
> ```

One issue with this approach in general is that the `panic` message can't display the error because `fmt` isn't `const`, so you can only give a generic message.

However, since there's a very limited set of possible in this case, we can do:

```rust
match Layout::from_size_align(BLOCK_SIZE, BLOCK_SIZE) {
    Ok(layout) => layout,
    Err(_) => panic!("{}",
        if BLOCK_SIZE == 0 {
            "Stream block size must not be zero"
        } else if !BLOCK_SIZE.is_power_of_two() {
            "Stream block size must be a power of two"
        } else if BLOCK_SIZE > (isize::MAX as usize) {
            "Stream block size must be within isize::MAX"
        } else {
            "Unexpected layout error"
        }
    )
}
```

Reply:
> With this PR merged we'll require at least 1.51 for const generics, which already exceeds the 1.48 in bullseye. I suspect that our MSRV may already be above that in practice. I've been building with 1.63 for quite a while, which is what's going to be in bookworm. I'm not sure we actually require it for anything yet, but I've not been checking that. How do the versions look on the other distros you looked at - what would be a sensible MSRV at this point?

Recent versions of Ubuntu/Fedora ship at least 1.65, I think the oldest one worth worrying about is 1.59 in openSUSE Leap 15.4. There's a good overall summary here: https://repology.org/project/rust/badges

Right now we still build successfully with 1.47 (https://github.com/miek/packetry/actions/runs/4447312275) so I think setting an MSRV of 1.57 for const_panic + const generics would be fine, and there's room to bump it up if we find something else we need.

Reply:
Cool!

I've squashed the `from_size_align` change into the commit that adds the const generics.

I've added a commit on the end which documents our MSRV as 1.57, and adds 1.57 to the CI matrix. Conveniently, Rust 1.56 added a `rust-version` field to `Cargo.toml`, so it's formally specified there there as well.